Admin Router: Prevent forwarding dcos-* cookies to upstreams

vespian opened the pull request:
## High Level Description

This PR introduces removing `dcos-*` cookies for upstream requests.

For technical motivation - please check the linked Jira issue
For technical implementation details - please check the description of the function `util.clear_dcos_cookies` in `lib/util.lua` file

## Related Issues

  - https://jira.mesosphere.com/browse/DCOS-12015 `Admin Router: remove auth cookie after translating it into Authorization header`

## Related PRs:

EE PR: https://github.com/mesosphere/dcos-enterprise/pull/1298

vespian commented:
@mesosphere-mergebot label ready for review

vespian commented:
@mhrabovcin @gpaul please have a look

mhrabovcin commented on packages/adminrouter/extra/src/test-harness/tests/test_generic.config.yml:
```diff
@@ -204,6 +206,7 @@ endpoint_tests:
   - tests:
       are_upstream_req_headers_ok:
         enabled: true
+        skip_authcookie_filtering_test: false
```
Can we have this option `false` by default?

vespian replied:
During the design phase of the syntax for generic tests YAML, I was considering having "silent defaults", but eventually decided against it and this is a general rule for all YAML-based tests. The reason is that we want/need people unfamiliar with AR repo to contribute. Being explicit about what is enabled and what is not is, I think, helpful as they do not need to browse through docs and/or wonder why their copy-pasted test does not work. 

In this particular case, I suppose, it is even better as it directly states the fact that auth-cookie passing test is performed unless explicitly disabled.

This is not written in stone, esp. as the number of generic tests increases and the amount of extra information grows. I do not think this is the moment though.

Happy to hear your thoughts on that.

jgehrcke commented:
Thanks, Pawel.

> as the generic tests syntax has changed and the documentation needs to be updated

Can you please treat the docs change as an independent task and -- with that -- push this PR towards a reviewable state?

I wonder why we do not hook in directly into [this](https://github.com/dcos/dcos/blob/06b089ac266b100503b0b0c71c159b429789e02f/packages/adminrouter/extra/src/lib/auth/common.lua#L69) code path. Note that the goal is to get rid of specifically the `dcos-acs-auth-cookie`. We do not care too much about the info cookie.

If the cookie filter stays implemented outside of the auth module I wonder if we should still trigger the filter logic conditionally from within the code I have pointed towards so that we don't unnecessarily inspect the cookies twice. Was there no easy way with having just one cookie jar? Thanks for clarification, I think addressing these points helps reviewers understanding the choices you have made.

mhrabovcin commented on packages/adminrouter/extra/src/test-harness/tests/test_generic.config.yml:
```diff
@@ -374,9 +393,22 @@ endpoint_tests:
   - tests:
       are_upstream_req_headers_ok:
         enabled: true
+        skip_authcookie_filtering_test: True
```
Should this be lowercase `true`?

mhrabovcin commented on packages/adminrouter/extra/src/test-harness/tests/test_generic.config.yml:
```diff
@@ -620,6 +661,12 @@ endpoint_tests:
   - tests:
+      are_upstream_req_headers_ok:
+        enabled: true
+        skip_authcookie_filtering_test: True
```
Lowercase `true`?

mhrabovcin commented on packages/adminrouter/extra/src/test-harness/modules/generic_test_code/generalised_tests.py:
```diff
@@ -413,23 +484,16 @@ def test_if_upstream_headers_are_correct(
-            assert_headers=headers_present,
-            assert_headers_absent=headers_absent,
+            jwt_forwarded_test,
+            skip_authcookie_filtering_test,
```
Other invocation uses named arg format: `skip_authcookie_filtering_test=skip_authcookie_filtering_test`

vespian replied:
Fixed - now all of them use positional args. Thx!

vespian commented:
> Can you please treat the docs change as an independent task and -- with that -- push this PR towards a reviewable state?

https://jira.mesosphere.com/browse/DCOS_OSS-1548

I scheduled it for this/next sprint

> I wonder why we do not hook in directly into this code path. 

Few reasons:
a) because this code is not always called. Please note that `util.clear_dcos_cookies`  is i.e. called here:
* https://github.com/dcos/dcos/pull/1827/files#diff-04b4f1568a453a99cae518ab0e07d0baR102
* https://github.com/dcos/dcos/pull/1827/files#diff-04b4f1568a453a99cae518ab0e07d0baR113
* https://github.com/mesosphere/dcos-enterprise/pull/1298/files#diff-c9d8d4dcd20678c47adb4a2804bf2e28R18
* https://github.com/mesosphere/dcos-enterprise/pull/1298/files#diff-c9d8d4dcd20678c47adb4a2804bf2e28R81
* https://github.com/mesosphere/dcos-enterprise/pull/1298/files#diff-c9d8d4dcd20678c47adb4a2804bf2e28R157

I suppose there will be more places like that in the future. We want to be thorough and never let the cookie through, hence a separate function.

b) because we should rethink how we circulate JWT token in the system, esp. considering the fact that we currently have no way to revoke it (apart from re-keying *all* JWTs, but this is a different thing). Lots of DC/OS services receive token either via an HTTP header or through a cookie. Compromising one of such service would mean compromising DC/OS security in general. 

This is a part of the bigger effort, that started during Admin Router refactoring and is tracked in     `DCOS-17924 Admin Router: JWT should not be forwarded to services that do not need`. 

One day I would really like to have AR pass JWT only to components that need it and nowhere else :)

> Note that the goal is to get rid of specifically the dcos-acs-auth-cookie. We do not care too much about the info cookie.

One of the goals of this PR is to hide `dcos-acs-auth-cookie` cookie from upstreams so that they can not start relying on it, even acidentally. The question is whether `dcos-acs-info-cookie` is an interface that we support and other services can rely upon. I think it is not (please correct me if I am wrong), hence the filtering. This is not written in stone, and as soon as there is a use case for this cookie, I will be more than happy to re-enable it, but let's cross this bridge.... :)

> If the cookie filter stays implemented outside of the auth module I wonder if we should still trigger the filter logic conditionally from within the code I have pointed towards so that we don't unnecessarily inspect the cookies twice. Was there no easy way to having just one cookie jar? 

I gave it a deeper thought. There are few reasons:
* the `Cookie` header parser is IMO pretty efficient: https://github.com/cloudflare/lua-resty-cookie/blob/master/lib/resty/cookie.lua#L62-L91 My *guess* is that it's O(n) but would have to formally analyse it to be sure
* in theory, we can share it but remember that we already need to support overlaying EE on top of Open which makes the code a bit "non-direct" already. I would like to avoid making it even more complex without strong technical reason (e.g. unacceptable increase of CPU usage) but at for now, considering that Lua code is JITed, I do not think it is worth it.
* fun fact: we would have been able to use `ngx.var.cookie_dcos_acs_auth_cookie` variable if it wasn't for the hyphens: http://lists.w3.org/Archives/Public/ietf-http-wg/2016JulSep/0015.html and https://github.com/openresty/lua-nginx-module/issues/19#issuecomment-111686154 My *guess* is that NGINX parser is much more effiicent than any Lua, JITed or not.

> Thanks for the clarification, I think addressing these points helps reviewers understanding the choices you have made.

nw, I hope that I managed to clarify all the points.

gpaul commented on packages/adminrouter/extra/src/lib/util.lua:
```diff
@@ -1,9 +1,65 @@
+
+    for k, v in pairs(cookies) do
+        if k ~= "dcos-acs-auth-cookie" and k ~= "dcos-acs-info-cookie" then
+            filtered_cookies[#filtered_cookies+1] = k .. "=" .. v
```
wow lua

gpaul commented on packages/adminrouter/extra/src/test-harness/modules/generic_test_code/common.py:
```diff
@@ -97,6 +98,67 @@ def generic_response_headers_verify_test(
+        headers (dict): headers dict that contains JWT and cookies. The auth
+            data it contains must be valid.
+        path (str): path for which request should be made
+        cookies_to_send (dic): dictionary containing all the cookies that should
```
dict

gpaul commented on packages/adminrouter/extra/src/test-harness/modules/generic_test_code/common.py:
```diff
@@ -97,6 +98,67 @@ def generic_response_headers_verify_test(
+            be send in the request
+        assert_cookies_present (dict): cookies to test where key is the
+            asserted cookie name and value is expected value of the cookie
+        assert_cookies_absent (dict): cookies that *MUST NOT* be present in the
```
This actually expects a `list()` or `set()` at the moment, which is great, the docstring just needs to get updated.

gpaul commented on packages/adminrouter/extra/src/test-harness/modules/generic_test_code/common.py:
```diff
@@ -97,6 +98,67 @@ def generic_response_headers_verify_test(
+
+    if len(cookie_headers) == 1:
+        jar = cookies.SimpleCookie()
+        # It is a list containing a single tuple, we need the second element of
```
:question: 
Nice comment, consider mentioning the tuple fields: "a single tuple `(name, value)`" if that is in fact what the tuple fields are

vespian replied:
```
# It is a list containing a single tuple (`header name`, `header value`),
# we need the second element of it - the value of the header:
```

gpaul commented on packages/adminrouter/extra/src/test-harness/modules/generic_test_code/common.py:
```diff
@@ -97,6 +98,67 @@ def generic_response_headers_verify_test(
+
+    # Let's make sure that we got not more than one 'Cookie' header:
+    # https://tools.ietf.org/html/rfc6265#section-5.4
+    cookie_headers = []
```
What do you think about using presence or `None` instead of  instead of using a list that has length 0 or 1?

vespian replied:
This list can have more than one element and this is an error. RFC forbids sending more than one `Cookie` header, but the tests just in case check for that before actually verifying cookies. If we switch to `None/Present`, we will miss the duplicated `Cookie` header scenario.

gpaul commented on packages/adminrouter/extra/src/test-harness/tests/test_generic.config.yml:
```diff
@@ -348,9 +366,21 @@ endpoint_tests:
 ######### /system/v1/logs endpoint
   - tests:
       are_upstream_req_headers_ok:
+        skip_authcookie_filtering_test: True
```
not sure if it matters but the false and true values don't have the same case

vespian replied:
uhhh... somehow I missed that, @mhrabovcin also pointed it out. I will grep the tests once more and try to fix all of them now.

jgehrcke commented:
> Few reasons:
> a) because this code is not always called. Please note that util.clear_dcos_cookies is i.e. called here:

(a) is concise and convinces me! I was not really able to parse the relation to my question from the remaining point(s?), but that is also not required anymore.

jgehrcke commented on packages/adminrouter/extra/src/test-harness/modules/generic_test_code/common.py:
```diff
@@ -97,6 +98,67 @@ def generic_response_headers_verify_test(
+
+    Arguments:
+        ar: Admin Router object, an instance of runner.(ee|open).Nginx
+        headers (dict): headers dict that contains JWT and cookies. The auth
```
Please replace "JWT" with what you actually mean. It's probably "the DC/OS authentication token".

jgehrcke commented on packages/adminrouter/extra/src/test-harness/modules/generic_test_code/generalised_tests.py:
```diff
@@ -305,6 +311,71 @@ def _testdata_to_redirect_testdata(tests_config, node_type):
+    # the dcos-acs-auth-cookie cookie is simply the DC/OS authentication token:
+    jar['dcos-acs-auth-cookie'] = valid_user_header['Authorization']
+    # the dcos-acs-info-cookie is base64-encoded data for dc/os UI,
+    # currently set to default user and password
```
> currently set to default user and password

What does that mean, and where does this come from? :)

The info cookie is a cookie accessible by JavaScript code (as opposed to the auth cookie) and just contains the `uid` of the user that is currently logged in, so that the UI can display it.

vespian replied:
> currently set to default user and password
Yep, this needs to be fixed :/

Will take care of it while working on https://jira.mesosphere.com/browse/DCOS_OSS-1572

jgehrcke commented on packages/adminrouter/extra/src/test-harness/modules/generic_test_code/generalised_tests.py:
```diff
@@ -305,6 +311,71 @@ def _testdata_to_redirect_testdata(tests_config, node_type):
+        'hbHNlfQ==')
+
+    # First case - cookie contains ONLY dcos-specific cookies, `Cookie` header
+    # should not be send
```
s/send/sent

vespian replied:
see above comment.

cdaw commented:
Merged in Train 208, please close.

vespian commented:
@cdaw Train 208 is still open/unmerged :)

cdaw commented:
sorry about that ... i thought we merged it ...
